src/software/thunderscope/robot_diagnostics/robot_error_log.py

```diff
@@ -0,0 +1,196 @@
+        # which prevents spamming the same battery warning
+        # set back to False if battery is back above warning level
+        self.low_battery_log_disabled: Dict[int, bool] = {}
+
+        # when the robot has an error code, its added to tne list keyed to the robot id
+        # which prevents spamming the same error code log
+        # list set back to empty if no error code
+        self.error_code_log_disabled: Dict[int, List[ErrorCode]] = {}
+
+        self.layout = QVBoxLayout()
+        self.layout.setAlignment(Qt.AlignmentFlag.AlignTop)
+
+        self.error_log_messages = []
+
+        # for a QScrollArea, widgets cannot be added to it directly
+        # doing so causes no scrolling to happen, and all the components get smaller
+        # instead, widgets are added to the layout which is set for a container
+        # the container is set as the current QScrollArea's widget
+        self.container = QFrame(self)
+        self.container.setLayout(self.layout)
+        self.setWidget(self.container)
+        self.setWidgetResizable(True)
+
+    def refresh(self) -> None:
+        """
+        Refreshes the widget's graphics
+
+        Gets the RobotCrash and RobotStatus messages from the buffers and adds
+        the corresponding log messages if needed
+
+        Updates the time since of the existing messages
+        """
+        # update time since for all existing logs
+        for error_message in self.error_log_messages:
+            error_message.update_time()
+
+        self.__refresh_fatal_logs()
+
+        self.__refresh_robot_crash()
+
+        # get a RobotStatus message
+        robot_status = self.robot_status_buffer.get(block=False, return_cached=False)
+
+        # empty out the buffer to get statuses for all current robots at once
+        while robot_status is not None:
+
+            self.__refresh_robot_status_error_code(robot_status)
+            self.__refresh_robot_status_battery_voltage(robot_status)
+
+            # get the next robot status and loop again
+            robot_status = self.robot_status_buffer.get(
+                block=False, return_cached=False
+            )
+
+    def __refresh_fatal_logs(self) -> None:
+        """
+        Gets a robot log from the buffer
+        If fatal or contract and last log is older than timeout, adds a log to the widget and
+        Records new last log time to prevent spamming
+        """
+        # get the robot log message
+        robot_log = self.robot_log_buffer.get(block=False, return_cached=False)
+
+        if robot_log is not None:
+            if (
+                robot_log.log_level == LogLevel.FATAL
+                or robot_log.log_level == LogLevel.CONTRACT
+            ):
+
+                if (
+                    robot_log.robot_id not in self.robot_last_fatal_time_s
+                    or time.time() - self.robot_last_fatal_time_s[robot_log.robot_id]
+                    > ROBOT_FATAL_TIMEOUT_S
```

Comment:
Do we really need two separate timeout constants?

This PR merged ROBOT_FATAL and ROBOT_CRASH:
https://github.com/UBC-Thunderbots/Software/pull/3038